File: web/src/test/java/hu/bsstudio/robonaut/speed/JuniorSpeedRaceResultHandlerTest.java
package hu.bsstudio.robonaut.speed;

import static org.mockito.Mockito.when;
import static org.mockito.MockitoAnnotations.openMocks;

import hu.bsstudio.robonaut.race.speed.SpeedRaceService;
import hu.bsstudio.robonaut.race.speed.model.SpeedRaceResult;
import hu.bsstudio.robonaut.team.model.DetailedTeam;
import java.util.Collections;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.springframework.test.web.reactive.server.WebTestClient;
import org.springframework.web.reactive.function.server.RouterFunctions;
import reactor.core.publisher.Mono;

final class JuniorSpeedRaceResultHandlerTest {

    @Mock
    private SpeedRaceService mockService;

    private WebTestClient webTestClient;

    @BeforeEach
    void setUp() {
        openMocks(this);
        final var underTest = new JuniorSpeedRaceResultHandler(mockService);
        final var routerFunction = RouterFunctions.route()
            .POST("/test", underTest).build();
        this.webTestClient = WebTestClient.bindToRouterFunction(routerFunction).build();
    }

    @Test
    void shouldReturnDetailedTeamWithOkStatus() {
        final var speedRaceResult = new SpeedRaceResult(0, 0, 0, Collections.emptyList());
        final var detailedTeam = DetailedTeam.builder().build();
        when(mockService.updateSpeedRaceJunior(speedRaceResult))
            .thenReturn(Mono.just(detailedTeam));

        webTestClient.post().uri("/test").bodyValue(speedRaceResult).exchange()
            .expectStatus().isOk()
            .expectBody(DetailedTeam.class).isEqualTo(detailedTeam);
    }
}

File: server/web/src/test/kotlin/hu/bsstudio/robonaut/speed/JuniorSpeedRaceResultHandlerTest.kt
package hu.bsstudio.robonaut.speed

import hu.bsstudio.robonaut.race.speed.SpeedRaceService
import hu.bsstudio.robonaut.race.speed.model.SpeedRaceResult
import hu.bsstudio.robonaut.team.model.DetailedTeam
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.test.web.reactive.server.WebTestClient
import org.springframework.test.web.reactive.server.expectBody
import org.springframework.web.reactive.function.server.RouterFunctions
import reactor.core.publisher.Mono

internal class JuniorSpeedRaceResultHandlerTest {
  @MockK
  private lateinit var mockService: SpeedRaceService
  private lateinit var webTestClient: WebTestClient

  @BeforeEach
  internal fun setUp() {
    MockKAnnotations.init(this)
    val underTest = JuniorSpeedRaceResultHandler(mockService)
    val routerFunction =
      RouterFunctions
        .route()
        .POST("/test", underTest)
        .build()
    webTestClient = WebTestClient.bindToRouterFunction(routerFunction).build()
  }

  @Test
  internal fun `should return DetailedTeam with OK status`() {
    val speedRaceResult = SpeedRaceResult(0, 0, 0, emptyList())
    val detailedTeam = DetailedTeam()
    every { mockService.updateSpeedRaceJunior(speedRaceResult) } returns Mono.just(detailedTeam)

    webTestClient
      .post()
      .uri("/test")
      .bodyValue(speedRaceResult)
      .exchange()
      .expectStatus()
      .isOk
      .expectBody<DetailedTeam>()
      .isEqualTo(detailedTeam)
  }
}
